### PR TITLE
Add support for CUDA Python

### PR DESCRIPTION
**Pull request by kmaehashi**
This PR rewrites the CuPy backend using [CUDA Python](https://github.com/NVIDIA/cuda-python). You will need to opt-in by `export CUPY_USE_CUDA_PYTHON=1` to build CuPy using CUDA Python. By adopting CUDA Python, we can avoid directly linking to CUDA libraries at build time.

Benefits of adopting CUDA Python are:

* Build-time link to CUDA Driver (`libcuda.so`) can be removed (c.f. https://github.com/cupy/cupy/pull/1781).
* CUDA libraries can be lazy-loaded: faster and smaller memory footprint when doing `import cupy` (c.f. https://github.com/cupy/cupy/issues/5557).
* When all CUDA Toolkit libraries are supported in CUDA Python:
    * The number of wheels to release can be reduced by utilizing CUDA Forward Compatibility, i.e., one release for each CUDA major version.
    * No longer need to maintain stub headers (which is currently needed to absorb CUDA version difference and build CuPy on RTD without CUDA).
    * CuPy can be imported without CUDA.

Note that the following components call CUDA Toolkit APIs from C/C++ layer (which bypasses CUDA Python):

* `cupy._core.cub` (via CUB header): Runtime
* `cupy._core.jitify` (via Jitify header): Runtime
* `cupy.cuda.thrust` (via Thrust header): Runtime
* `cupy.random._generator_api` (via cuRAND header): Runtime, cuRAND
* `cupy_backends.cuda.libs.profiler`: Runtime
    * This module calls `cudaProfilerInitialize` but is unsupported by CUDA Python, as the API is a deprecated one

We are planning to build CuPy v10 wheels using CUDA Python once binary distributions of CUDA Python become available.

cc/ @mmccarty @vzhurba01

### CODE REVIEW

**Comment by kmaehashi** [comment]
pfnCI, test this please.

File: cupy_backends/cuda/api/_driver_typedef.pxi [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    # For Texture Reference
+    ctypedef void* Array 'CUarray_st*'  # = cupy.cuda.runtime.Array
+    ctypedef int Array_format 'CUarray_format'
+    ctypedef struct Array_desc 'CUDA_ARRAY_DESCRIPTOR':
+        Array_format Format
+        size_t Height
+        unsigned int NumChannels
+        size_t Width
+    ctypedef int Address_mode 'CUaddress_mode'
+    ctypedef int Filter_mode 'CUfilter_mode'
```
Comment (leofang) [inline cupy_backends/cuda/api/_driver_typedef.pxi:31]: Note to self: Remove the texture reference support as it's deprecated and unsupported by CUDA Python.

File: cupy_backends/cuda/api/driver.pyx [hunk 2]
```diff
@@ -15,89 +15,23 @@ cimport cython  # NOQA
...
-        block2shmem, size_t dynamicSMemSize, int blockSizeLimit)
+IF CUPY_USE_CUDA_PYTHON:
+    from cuda.ccuda cimport *
+ELSE:
+    include '_driver_extern.pxi'
 
+cdef extern from '../../cupy_backend.h' nogil:
     # Build-time version
-    int CUDA_VERSION
     int HIP_VERSION
```
Comment (leofang) [inline cupy_backends/cuda/api/driver.pyx:30]: Can we leave a comment here? Something like
```suggestion
    # Note: CUDA_VERSION is defined either in CUDA Python or _driver_extern.pxi
    int HIP_VERSION
```

File: cupy_backends/cuda/api/driver.pyx [hunk 3]
```diff
@@ -15,89 +15,23 @@ cimport cython  # NOQA
...
 
+# Provide access to constants from Python.
+from cupy_backends.cuda.api._driver_enum import *
```
Comment (leofang) [inline cupy_backends/cuda/api/driver.pyx:33]: In the case of CUDA Python, can't we utilize the enums exposed to Python from `cudapython/cuda.pyx`? Though the current approach is probably a lot easier to manage.

File: cupy_backends/cuda/api/driver.pyx [hunk 4]
```diff
@@ -148,6 +82,10 @@ cpdef get_build_version():
         return 0
 
 
+cpdef bint _is_cuda_python():
+    return CUPY_USE_CUDA_PYTHON
```
Comment (leofang) [inline cupy_backends/cuda/api/driver.pyx:86]: Just a thought, haven't figured it out completely: Maybe this should live in the "Constants (CuPy)" section in `runtime.pyx`, sitting side by side with `_is_hip_environment`?

File: cupyx/_runtime.py [hunk 5]
```diff
@@ -149,7 +149,11 @@ def __init__(self, *, full=True):
         else:
             self.nvcc_path = cupy._environment.get_hipcc_path()
 
-        self.cuda_build_version = cupy.cuda.driver.get_build_version()
+        driver_build_version = cupy.cuda.driver.get_build_version()
+        if cupy.cuda.driver._is_cuda_python():
```
Comment (leofang) [inline cupyx/_runtime.py:153]: Ah I see `_is_cuda_python()` is used here.

**Review by leofang — DISMISSED** [review]
LGTM except for a few questions! This is great work @kmaehashi!

File: cupy_backends/cuda/api/_driver_typedef.pxi (hunk 1 above)
Comment (kmaehashi) [reply cupy_backends/cuda/api/_driver_typedef.pxi:31]: It's deprecated but seems `CUtexref` exists in CUDA Python. (maybe I'm missing something?)

File: cupy_backends/cuda/api/driver.pyx (hunk 3 above)
Comment (kmaehashi) [reply cupy_backends/cuda/api/driver.pyx:34]: Thanks, applied!

File: cupy_backends/cuda/api/driver.pyx (hunk 2 above)
Comment (kmaehashi) [reply cupy_backends/cuda/api/driver.pyx:31]: Thanks, applied!

File: cupy_backends/cuda/api/driver.pyx (hunk 3 above)
Comment (kmaehashi) [reply cupy_backends/cuda/api/driver.pyx:34]: That's possible but we have to maintain `_driver_enum` for HIP anyways.

File: cupy_backends/cuda/api/driver.pyx (hunk 4 above)
Comment (kmaehashi) [reply cupy_backends/cuda/api/driver.pyx:87]: I'm thinking it's better to place this here considering `get_build_version()` returns the version of CUDA Python.

**Comment by kmaehashi** [comment]
pfnCI, test this please.

File: cupy_backends/cuda/api/_driver_typedef.pxi (hunk 1 above)
Comment (leofang) [reply cupy_backends/cuda/api/_driver_typedef.pxi:31]: nvm I read it wrong, sorry!

**Comment by kmaehashi** [comment]
Added support CI with CUDA Python.
https://ci.preferred.jp/cupy.linux.cuda11x-cuda-python/77003/

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Comment by kmaehashi** [comment]
All tests passed :tada:

**Comment by mmccarty** [comment]
Thanks everyone! I'm so excited to see the adoption of CUDA Python in CuPy!

**Review by emcastillo — APPROVED** [review]
LGTM!
I don't have concerns with this PR, I think it is better to merge it now and address all the issues that may happen incrementally

**Pull request merged**